### PR TITLE
Multi Output With Missing Values

### PR DESCRIPTION
**Pull request by samcarlos**
Including ability to have NA's (as -1's) in response variables for multi output models.

### CODE REVIEW

File: muffnn/core.py [hunk 1]
```diff
@@ -7,6 +7,52 @@
 from tempfile import NamedTemporaryFile
 import tensorflow as tf
 
+def affine_lasso(input_tensor, output_size, bias=True, bias_start=0.0,
```
Comment (beckermr) [inline muffnn/core.py]: You don't need this function. You can just use the `affine` function with `bias=False`.

File: muffnn/mlp/base.py [hunk 2]
```diff
@@ -112,6 +112,7 @@ def partial_fit(self, X, y, monitor=None, **kwargs):
         if self._is_fitted:
             y = self._transform_targets(y)
         else:
+            self._checks = []
```
Comment (beckermr) [inline muffnn/mlp/base.py]: We used these checks for debugging. We should def remove them now.

File: muffnn/mlp/base.py [hunk 3]
```diff
@@ -134,6 +135,8 @@ def partial_fit(self, X, y, monitor=None, **kwargs):
 
                 # Train model parameters.
                 self._session.run(tf.initialize_all_variables())
+                self._checks.append(tf.check_numerics(self._obj_func, 'objfunc is bad!'))
```
Comment (beckermr) [inline muffnn/mlp/base.py]: Ditto here for removing checks.

Comment (beckermr) [inline muffnn/mlp/base.py]: And remove the extra line below.

File: muffnn/mlp/base.py [hunk 4]
```diff
@@ -146,15 +149,29 @@ def partial_fit(self, X, y, monitor=None, **kwargs):
...
+                    ce = vals[4]
                     _LOGGER.debug("objective: %.4f, epoch: %d, idx: %d",
                                   obj_val, epoch, start_idx)
-
+                    print(ce)
+                    print(selections)
+                    print(selections.shape)
+                    print(selections.sum())
+                    print(obj_val)
+                    print(yfin)
```
Comment (beckermr) [inline muffnn/mlp/base.py]: All of the extra print statements here need to be removed.

File: muffnn/mlp/base.py [hunk 5]
```diff
@@ -146,15 +149,29 @@ def partial_fit(self, X, y, monitor=None, **kwargs):
...
-                        [self._obj_func, self._train_step],
+                                                     yb)
+
+                    vals = self._session.run(
+                        [self._obj_func, self._train_step_step, self._selections, self._yfinite, self.ce] + self._checks,
                         feed_dict=feed_dict)
+                    obj_val = vals[0]
+                    selections = vals[2]
+                    yfin = vals[3]
+                    ce = vals[4]
```
Comment (beckermr) [inline muffnn/mlp/base.py]: All of the code above needs to be changed back to the original form. This is leftover code from our hacking on this.

File: muffnn/mlp/base.py [hunk 6]
```diff
@@ -146,15 +149,29 @@ def partial_fit(self, X, y, monitor=None, **kwargs):
             for epoch in range(self.n_epochs):
                 self._random_state.shuffle(indices)
                 for start_idx in range(0, n_examples, self.batch_size):
+                    self.zeros=tf.constant(0, dtype=tf.float32,shape=[self.batch_size,self.n_classes_])
```
Comment (beckermr) [inline muffnn/mlp/base.py]: I don't think this variable is used here, right?

File: muffnn/mlp/base.py [hunk 7]
```diff
@@ -246,14 +265,20 @@ def _set_up_graph(self):
                 if self.keep_prob != 1.0:
                     t = tf.nn.dropout(t, keep_prob=self._keep_prob)
                 t = affine(t, layer_sz, scope='layer_%d' % i)
+                self._checks.append(tf.check_numerics(t, 'layer_%d' % i))
```
Comment (beckermr) [inline muffnn/mlp/base.py]: Remove this.

File: muffnn/mlp/base.py [hunk 8]
```diff
@@ -246,14 +265,20 @@ def _set_up_graph(self):
...
         t = self._init_model_output(t)
         self._init_model_objective_fn(t)
 
-        self._train_step = tf.train.AdamOptimizer().minimize(self._obj_func)
+        self._train_step = tf.train.AdamOptimizer()
+        self._grads_and_vars = self._train_step.compute_gradients(self._obj_func, tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES))
+        # for a, b in self._grads_and_vars:
+        #     self._checks.append(tf.check_numerics(a, a.name))
+        #     self._checks.append(tf.check_numerics(b, b.name))
+        self._train_step_step = self._train_step.apply_gradients(self._grads_and_vars)
```
Comment (beckermr) [inline muffnn/mlp/base.py]: We only split the train step up for debugging. You should put all of this back to the original form.

File: muffnn/mlp/mlp_classifier.py [hunk 9]
```diff
@@ -105,28 +105,35 @@ def _init_model_output(self, t):
...
                 t, self.input_targets_)
         else:
             cross_entropy = tf.nn.sigmoid_cross_entropy_with_logits(
                 t, tf.cast(self.input_targets_, np.float32))
-        self._obj_func = tf.reduce_mean(cross_entropy)
+        y_finite=tf.equal(self.input_targets_, -1)
+        self._selections=tf.select(y_finite, self.zeros, cross_entropy )
+        self._yfinite=y_finite
+        self.ce=cross_entropy
+        self._obj_func =  tf.reduce_sum(tf.select(y_finite,  self.zeros,cross_entropy ))
```
Comment (beckermr) [inline muffnn/mlp/mlp_classifier.py]: We should build the zeros on the fly here, not in the training function above (which is actually part of the base class). You can use tf.zeros_like to get a Tensor of zeros that is the same shape as the input targets.

Remove the extra attributes we set during debugging, like `ce`, `selections`, `_yfinite`.

File: test_na_muffn_iris.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+from muffnn import MLPClassifier
```
Comment (beckermr) [inline test_na_muffn_iris.py]: You need to write a real set of tests in the tests directory of the mlp module. Put these tests there in the same format as the ones that are there and then delete this file.

File: muffnn/mlp/mlp_classifier.py [hunk 11]
```diff
@@ -167,7 +174,11 @@ def _is_multilabel(self, y):
         Return whether the given target array corresponds to a multilabel
         problem.
         """
-        target_type = type_of_target(y)
+        temp_y=y.copy()
```
Comment (samcarlos) [inline muffnn/mlp/mlp_classifier.py]: I think this step is still necessary because 'type_of_target' can only have two unique values. Having 3 (1,0, -1) will not result in multi class indicator outputs: https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/multiclass.py#L198

Comment (beckermr) [inline muffnn/mlp/mlp_classifier.py]: Yep that sounds right.

File: muffnn/mlp/base.py [hunk 12]
```diff
@@ -133,9 +133,9 @@ def partial_fit(self, X, y, monitor=None, **kwargs):
                 self._build_tf_graph()
 
                 # Train model parameters.
-                self._session.run(tf.initialize_all_variables())
-
-            # Set an attributed to mark this as at least partially fitted.
+                self._session.run(tf.initialize_all_variables())\
+                    # Set an attributed to mark this as at least partially
```
Comment (beckermr) [inline muffnn/mlp/base.py]: What is going on with the comment spacing here?

File: muffnn/mlp/mlp_classifier.py [hunk 13]
```diff
@@ -105,28 +105,35 @@ def _init_model_output(self, t):
...
             cross_entropy = tf.nn.sparse_softmax_cross_entropy_with_logits(
                 t, self.input_targets_)
         else:
             cross_entropy = tf.nn.sigmoid_cross_entropy_with_logits(
                 t, tf.cast(self.input_targets_, np.float32))
-        self._obj_func = tf.reduce_mean(cross_entropy)
+        y_finite = tf.equal(self.input_targets_, -1)
+        self.zeros = tf.constant(
+            0, dtype=tf.float32, shape=[
+                self.batch_size, self.n_classes_])
```
Comment (beckermr) [inline muffnn/mlp/mlp_classifier.py]: You want `tf.zeros_like` here to make sure that `self.zeros` is the same shape as the input targets. Also, you should use `self._zeros` since we don't want this attribute exposed to the user.

File: muffnn/mlp/mlp_classifier.py [hunk 14]
```diff
@@ -105,28 +105,35 @@ def _init_model_output(self, t):
 
         if self.multilabel_:
             self.input_targets_ = \
-                tf.placeholder(tf.int64, [None, self.n_classes_], "labels")
+                tf.placeholder(tf.float32, [None, self.n_classes_], "targets")
             self.output_layer_ = tf.nn.sigmoid(t)
+            self.zeros = tf.zeros_like(self.output_layer_)
```
Comment (beckermr) [inline muffnn/mlp/mlp_classifier.py]: You don't need this line to define `self.zeros`.

File: muffnn/mlp/mlp_classifier.py (hunk 13 above)
Comment (beckermr) [inline muffnn/mlp/mlp_classifier.py]: We should only apply the select for multilabel. Move the null handling into the multilabel part of the if else and then just put back the reduce_mean of the cross entropy in the other cases.

File: muffnn/mlp/mlp_classifier.py [hunk 15]
```diff
@@ -105,28 +105,35 @@ def _init_model_output(self, t):
...
                 t, self.input_targets_)
         else:
             cross_entropy = tf.nn.sigmoid_cross_entropy_with_logits(
                 t, tf.cast(self.input_targets_, np.float32))
-        self._obj_func = tf.reduce_mean(cross_entropy)
+        y_finite = tf.equal(self.input_targets_, -1)
+        self.zeros = tf.constant(
+            0, dtype=tf.float32, shape=[
+                self.batch_size, self.n_classes_])
+        self._obj_func = tf.reduce_sum(
```
Comment (beckermr) [inline muffnn/mlp/mlp_classifier.py]: I think this should be a reduce_mean, right?

File: muffnn/mlp/mlp_classifier.py [hunk 16]
```diff
@@ -105,8 +105,10 @@ def _init_model_output(self, t):
 
         if self.multilabel_:
             self.input_targets_ = \
-                tf.placeholder(tf.int64, [None, self.n_classes_], "labels")
+                tf.placeholder(tf.int64, [None, self.n_classes_], "targets")
             self.output_layer_ = tf.nn.sigmoid(t)
+            self._zeros = tf.zeros_like(self.output_layer_)
+
```
Comment (beckermr) [inline muffnn/mlp/mlp_classifier.py]: Remove this extra line.

File: muffnn/mlp/base.py [hunk 17]
```diff
@@ -171,7 +171,6 @@ def _check_inputs(self, X, y):
         # Check that the input X is an array or sparse matrix.
         # Convert to CSR if it's in another sparse format.
         X, y = check_X_y(X, y, accept_sparse='csr', multi_output=True)
-
```
Comment (beckermr) [inline muffnn/mlp/base.py]: Add back this extra line.

File: muffnn/mlp/mlp_classifier.py [hunk 18]
```diff
@@ -173,10 +173,7 @@ def _is_multilabel(self, y):
         problem.
         """
         temp_y = y.copy()
-        if np.isnan(temp_y).sum() > 0:
-            raise ValueError(
-                "Cannot have NA's in response variable. Please change to -1")
-        temp_y[(temp_y == -1)] = 1
+        temp_y[np.where(temp_y == -1)] = 1
```
Comment (beckermr) [inline muffnn/mlp/mlp_classifier.py]: BTW, you don't need the np.where call here.

```python
temp_y[temp_y == -1] = 1
```

should work fine.

File: muffnn/mlp/mlp_classifier.py [hunk 19]
```diff
@@ -167,7 +172,9 @@ def _is_multilabel(self, y):
         Return whether the given target array corresponds to a multilabel
         problem.
         """
-        target_type = type_of_target(y)
+        temp_y = y.copy()
+        temp_y[np.where(temp_y == -1)] = 1
```
Comment (beckermr) [inline muffnn/mlp/mlp_classifier.py]: We should change this line to

```python
temp_y[np.zeros_like(temp_y, dtype=bool) | (temp_y == -1)] = 1
```

This gives the right output

```python
In [57]: temp_y = np.array(['a', 'b', 'c'])

In [58]: np.zeros_like(temp_y, dtype=bool) | (temp_y == -1)
/Users/mbecker/miniconda3/envs/civis/bin/ipython:1: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  #!/Users/mbecker/miniconda3/envs/civis/bin/python
Out[58]: array([False, False, False], dtype=bool)

In [59]: temp_y = np.array([['a', 'b'], ['a', 'b']])

In [60]: np.zeros_like(temp_y, dtype=bool) | (temp_y == -1)
/Users/mbecker/miniconda3/envs/civis/bin/ipython:1: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  #!/Users/mbecker/miniconda3/envs/civis/bin/python
Out[60]: 
array([[False, False],
       [False, False]], dtype=bool)

In [61]: temp_y = np.array([0, -1, 1])

In [62]: np.zeros_like(temp_y, dtype=bool) | (temp_y == -1)
Out[62]: array([False,  True, False], dtype=bool)
```

File: muffnn/mlp/tests/test_mlp_classifier.py [hunk 20]
```diff
@@ -65,11 +68,31 @@ def check_multilabel_predictions(clf, X, y):
...
+    assert_equal(predicted.shape, y.shape)
+    predicted[is_na]=0
+    y[is_na]=0
+    assert_array_equal(predicted, y)
+
+    probabilities = clf.predict_proba(X)
+    assert_equal(probabilities.shape, y.shape)
+    assert_array_equal((probabilities >= 0.5).astype(np.int), predicted)
+    assert np.sum(np.abs(probabilities.sum(axis=1) - 1)) > 0.01,\
+        "multilabel classifier is outputting distributions"
```
Comment (beckermr) [inline muffnn/mlp/tests/test_mlp_classifier.py]: This message is wrong. Change the other one too.

File: muffnn/mlp/tests/test_mlp_classifier.py [hunk 21]
```diff
@@ -65,11 +68,31 @@ def check_multilabel_predictions(clf, X, y):
...
+def check_multilabel_predictions_na(clf, X, y):
+    predicted = clf.fit(X, y).predict(X)
+    #change nans to zero to 'mask' those values
+    is_na=(y==-1)
+    predicted[is_na]=0
+    y[is_na]=0
+
+    assert_equal(predicted.shape, y.shape)
+    predicted[is_na]=0
+    y[is_na]=0
```
Comment (beckermr) [inline muffnn/mlp/tests/test_mlp_classifier.py]: These two lines are duplicates.

File: muffnn/mlp/tests/test_mlp_classifier.py [hunk 22]
```diff
@@ -65,11 +68,31 @@ def check_multilabel_predictions(clf, X, y):
     assert np.sum(np.abs(probabilities.sum(axis=1) - 1)) > 0.01,\
         "multilabel classifier is outputting distributions"
 
+def check_multilabel_predictions_na(clf, X, y):
+    predicted = clf.fit(X, y).predict(X)
+    #change nans to zero to 'mask' those values
+    is_na=(y==-1)
+    predicted[is_na]=0
+    y[is_na]=0
```
Comment (beckermr) [inline muffnn/mlp/tests/test_mlp_classifier.py]: You might want a copy of y first.

File: muffnn/mlp/tests/test_mlp_classifier.py [hunk 23]
```diff
@@ -65,11 +68,31 @@ def check_multilabel_predictions(clf, X, y):
...
+
+    assert_equal(predicted.shape, y.shape)
+    predicted[is_na]=0
+    y[is_na]=0
+    assert_array_equal(predicted, y)
+
+    probabilities = clf.predict_proba(X)
+    assert_equal(probabilities.shape, y.shape)
+    assert_array_equal((probabilities >= 0.5).astype(np.int), predicted)
```
Comment (beckermr) [inline muffnn/mlp/tests/test_mlp_classifier.py]: Do you need to use the is_nan mask here on probabilities after converting to int?

File: muffnn/mlp/tests/test_mlp_classifier.py [hunk 24]
```diff
@@ -66,11 +73,46 @@ def check_multilabel_predictions(clf, X, y):
...
+    y_temp = y.copy()
+    y_temp[is_na] = 0
+
+    assert_equal(predicted.shape, y_temp.shape)
+    assert_array_equal(predicted, y_temp)
+
+    probabilities = clf.predict_proba(X)
+    probabilities[is_na] = 0
+    assert_equal(probabilities.shape, y_temp.shape)
+    assert_array_equal((probabilities >= 0.5).astype(np.int), predicted)
```
Comment (beckermr) [inline muffnn/mlp/tests/test_mlp_classifier.py]: Do the `probabilities` need to be masked here in this test?

File: test_na_muffn_iris.py (hunk 10 above)
Comment (beckermr) [inline test_na_muffn_iris.py]: Def make sure to remove this file.

File: muffnn/mlp/tests/test_mlp_classifier.py (hunk 24 above)
Comment (samcarlos) [inline muffnn/mlp/tests/test_mlp_classifier.py]: I think so. `Probabilities` is being compared to `predicted`, which is being masked.

Comment (beckermr) [inline muffnn/mlp/tests/test_mlp_classifier.py]: Ack. Yes of course I missed the na line here above. Sorry! Kudos!

**Comment by beckermr** [comment]
OK. Let's push to a branch in muffnn itself to get the tests to run. Then we can finish the PR. Note the branch here and close this PR after you are done.

**Comment by samcarlos** [comment]
PR changed to "Adding NA functionality for multilabel classification problems #25"
